kafka/producer.go

```diff
@@ -17,4 +17,170 @@
 
 package kafka
 
-// TODO
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"go.uber.org/zap"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kzap"
+
+	"github.com/elastic/apm-data/model"
+	"github.com/elastic/apm-queue/queuecontext"
+)
+
+// Encoder encodes a model.APMEvent to a []byte
+type Encoder interface {
+	// Encode accepts a model.APMEvent and returns the encoded representation.
+	Encode(model.APMEvent) ([]byte, error)
+}
+
+// RecordMutator mutates the record associated with the model.APMEvent
+type RecordMutator func(model.APMEvent, *kgo.Record) error
+
+// ProducerConfig holds configuration for publishing events to Kafka.
+type ProducerConfig struct {
+	// Broker holds the (host:port) address of the Kafka broker to which
+	// events should be published.
+	Broker string
+
+	// ClientID to use when connecting to Kafka. This is used for logging
+	// and client identification purposes.
+	ClientID string
+	// Version is the software version to use in the Kafka client. This is
+	// useful since it shows up in Kafka metrics and logs.
+	Version string
+
+	// Logger is used for logging producer errors.
+	Logger *zap.Logger
+
+	// Encoder holds an encoding.Encoder for encoding events.
+	Encoder Encoder
+
+	// Sync can be used to indicate whether production should be synchronous.
+	Sync bool
+
+	// Mutators holds the list of RecordMutator applied to all the records sent
+	// by the producer.
+	Mutators []RecordMutator
+}
+
+// Validate checks that cfg is valid, and returns an error otherwise.
+func (cfg ProducerConfig) Validate() error {
+	if cfg.Broker == "" {
+		return errors.New("broker cannot be empty")
+	}
+	if cfg.Logger == nil {
+		return errors.New("logger cannot be nil")
+	}
+	return nil
```

Comment:
I missed this in the previous review: Could you prefix the errors with the name of the package and use Go `1.20` semantics to be able to return a _multi-error_? Similar to what's being done here: https://github.com/elastic/apm-queue/blob/7e52ea6dc17dd6c60dbd6fd62651cc11c94997f2/kafka/consumer.go#L64-L85.

We're also missing a nil check on the cfg.Encoder. Could you add that as well, so caller's don't cause a panic?

Reply:
Thanks for mentioning this! Updated!